Add IntegratedLabelerSelector Class and Notebook

landoskape opened the pull request:
I really liked the scatterSelector functionality and wanted to use it in combination with the imagelabeler. So I combined them into a single integrated GUI. Now, you can draw a lasso around points embedded with UMAP and specifically label those ROIs with the labeler. So you can easily see what kinds of ROIs you have (with the image overlay) and then label a bunch of examples of those kinds of ROIs. 

The new class is in the helpers module - I had to convert the scatterSelector and ImageLabeler plotting to use matplotlib because this made it easier to get the two components talking to each other. I believe it's about just as fast (I think it could be improved slightly if it's of interest). 

I also added another example notebook in the classification in place of B1 (called B1alt) as well in case you want to switch to this. 

I've tried it out on windows only, let me know if it works on linux/mac/whatever you're using

![image](https://github.com/user-attachments/assets/8ad15ec8-8928-4d38-ba02-ac42ea38c613)

RichieHakim commented:
My man

landoskape commented:
enjoy it while it lasts

RichieHakim commented:
Ok, I gave it a go. It's awesome. I will definitely integrate it somehow. There are some issues I'm not sure how to address.
1. On colab it doesn't open:
```python
<ipython-input-17-465377b69ae6>:216: UserWarning: Error initializing image labeler: no display name and no $DISPLAY environment variable
  warnings.warn('Error initializing image labeler: ' + str(e))

---------------------------------------------------------------------------

TypeError                                 Traceback (most recent call last)

[<ipython-input-20-88cab3542db2>](https://localhost:8080/#) in <cell line: 21>()
     19 }
     20 
---> 21 print(f"Number of labeled ROIs: {len(labels['labels_list'])}")
     22 print(f"path to csv file: {labeler.path_csv}")

TypeError: object of type 'NoneType' has no len()
```

2. On my linux machine using both jupyter and vscode's notebook, it opens and the UI seems to work well, but it doesn't output labels:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[11], [line 19](vscode-notebook-cell:?execution_count=11&line=19)
     [11](vscode-notebook-cell:?execution_count=11&line=11)     labeler.run()
     [13](vscode-notebook-cell:?execution_count=11&line=13) labels = {
     [14](vscode-notebook-cell:?execution_count=11&line=14)     'labels_list': labeler.get_labels(kind='list'),
     [15](vscode-notebook-cell:?execution_count=11&line=15)     'labels_dict': labeler.get_labels(kind='dict'),
     [16](vscode-notebook-cell:?execution_count=11&line=16)     'labels_df': labeler.get_labels(kind='dataframe'),
     [17](vscode-notebook-cell:?execution_count=11&line=17) }
---> [19](vscode-notebook-cell:?execution_count=11&line=19) print(f"Number of labeled ROIs: {len(labels['labels_list'])}")
     [20](vscode-notebook-cell:?execution_count=11&line=20) print(f"path to csv file: {labeler.path_csv}")

TypeError: object of type 'NoneType' has no len()
```

Let me know how to proceed. Happy to get on a zoom to screen share / peer code.

landoskape commented:
## Colab
I just tested it in colab also, and it definitely doesn't work because colab doesn't have access to the display. There are apparently some shortcuts (like ``enable_custom_widget_manager()``), but I think they only work inline (I might be mistaken). Did you ever get your ImageLabeler class to work in colab? If so, how? If you think it's important to make it work in colab I think I can remove the tkinter elements and do it all in matplotlib, which will probably work as an inline widget. 

## Linux Behavior
I copied your ``get_labels`` method, which returns ``None`` when nothing is labeled. Have you tried to label some images first? If you get the same error I'll be surprised. 

If that solves it - then one solution would be to change that behavior of ``get_labels`` to return ``['None', ] * len(self.images)`` as you have when there are labels. I'd recommend changing it for the ``ImageLabeler`` class too in helpers.

RichieHakim commented:
I'm sorry for dragging my feet on this. Something something surgery thesis

I found what was blocking the assignment of labels on Ubuntu. It's this block in the 'classify' method:
```python
        ...
        # Prevent classify from running with any special keys
        if event.state != 8:
            return 
        ...
```

Do you think this is necessary?
I find that if I print the event.state just before these lines, I get: `event.state = 16`.

landoskape commented:
> I'm sorry for dragging my feet on this. Something something surgery thesis
> 
> I found what was blocking the assignment of labels on Ubuntu. It's this block in the 'classify' method:
> 
> ```python
>         ...
>         # Prevent classify from running with any special keys
>         if event.state != 8:
>             return 
>         ...
> ```
> 
> https://github.com/landoskape/ROICaT/blob/integrated-labeler/roicat/helpers.py#L3157
> 
> Do you think this is necessary? I find that if I print the event.state just before these lines, I get: `event.state = 16` when I do a normal keypress.

Ah - I see I guess I don't really understand the state assignment. I was trying to prevent a label from being "Ctrl-L" for example. I don't think it's necessary because this leads to the event.char being set to ``''`` which you've already accounted for! I've removed that line from the PR. 

> And don't worry about colab. tkinter generally doesn't work in colab I guess. I had to use bokeh to get the drawing classification working in colab, and it's not worth the effort.

Okay, yeah, it's too bad they aren't compatible. In case people try to open this integrated-labeler notebook in colab, I added a highlighted line to state that it doesn't work at the top of the notebook.

landoskape commented:
Any other thoughts? Seems mergeable? Let me know if you have other ideas / requests

RichieHakim commented:
@landoskape 
> Any other thoughts? Seems mergeable? Let me know if you have other ideas / requests

I was trying to figure out a way to merge what you wrote with one of the other notebooks, ideally the drawing classification one, but I couldn't think of a natural way to do it. I thought it would be cool if there was a way to label all the ROIs within a circled group in addition to the single ROI method; I think it'll be too much effort to do something elegant. I'll merge it with a couple small changes. 

1. There is a small redundancy in the notebook:

```python
labels = {
    'labels_list': labeler.get_labels(kind='list'),
    'labels_dict': labeler.get_labels(kind='dict'),
    'labels_df': labeler.get_labels(kind='dataframe'),
}

print(f"Number of labeled ROIs: {len(labels['labels_list'])}")
print(f"path to csv file: {labeler.path_csv}")

labels = {
    'list': roicat.util.JSON_List(labeler.get_labels(kind='list')),
    'df': labeler.get_labels(kind='dataframe'),
}

```

2. For printing the labeled ROI, print the repr to account for special characters:
```python
            print(f'Image {self._get_current_idx()}: {str(repr(label))}') if self._verbose else None
```

3. Change the name of the notebook. This is whatever and I'll defer to you if you want to name it 'alt' or anything else. Now the two labeling notebooks are B1a and B1b, for ex: `B1b_labeling_drawingAndInteractive.ipynb`

4. Update the readme with the new notebook names

RichieHakim commented:
I just pushed things onto your branch. Hopefully that is okay. You can revert if you want. If things look good to you I'll merge.

landoskape commented:
Looks good to me! I agree with your changes and points. 

Regarding the label all ROIs in a drawing area - I was thinking about that too, but I think that labeling ROIs examples directly is probably safer because you can just label good examples, then your classifier can do the rest of the work (the user might pick out details that don't show up in 2D UMAP space). 

Okay and I'm pretty sure the testing errors on ``test_integration`` are unrelated to this change.
